tt_metal/impl/device/device.cpp

```diff
@@ -915,99 +955,278 @@ void Device::update_workers_build_settings(std::vector<std::vector<std::tuple<tt
+                        compile_args[0] = settings.cb_start_address;
+                        compile_args[1] = settings.cb_log_page_size;
+                        compile_args[2] = settings.cb_pages;
+                        compile_args[3] = settings.consumer_semaphore_id;
+                        compile_args[4] = demux_sem++;
+                        compile_args[5] = dispatch_constants::DISPATCH_BUFFER_SIZE_BLOCKS;
+                        compile_args[6] = 0; //unused prefetch_sync_sem
+                        compile_args[7] = settings.command_queue_start_addr;
+                        compile_args[8] = settings.completion_queue_start_addr;
+                        compile_args[9] = settings.completion_queue_size;
+                        compile_args[10] = dispatch_constants::DISPATCH_BUFFER_BASE; // unused
+                        compile_args[11] = (1 << dispatch_constants::DISPATCH_BUFFER_LOG_PAGE_SIZE) * dispatch_constants::get(dispatch_core_type).dispatch_buffer_pages(); // unused
+                        compile_args[12] = 0; // unused: local ds semaphore
+                        compile_args[13] = 0; // unused: remote ds semaphore
+                        compile_args[14] = 0; // preamble size
+                        compile_args[15] = true,    // split_prefetcher
+                        compile_args[16] = NOC_XY_ENCODING(prefetch_physical_core.x, prefetch_physical_core.y),
+                        compile_args[17] = prefetch_h_settings.producer_semaphore_id, // sem_id on prefetch_h that dispatch_d is meant to increment, to resume sending of cmds post exec_buf stall
+                        compile_args[18] = dispatch_constants::get(dispatch_core_type).mux_buffer_pages(num_hw_cqs), // XXXX should this be mux pages?
+                        compile_args[19] = settings.num_compute_cores;
+                        compile_args[20] = 0; // unused: dispatch_d only
+                        compile_args[21] = 0; // unused: dispatch_d only
+                        compile_args[22] = 0; // unused: dispatch_d only
+                        compile_args[23] = 0; // unused: dispatch_d only
+                        compile_args[24] = 0;
+                        compile_args[25] = false; // is_dram_variant
+                        compile_args[26] = true; // is_host_variant
+
+                        dispatch_idx++;
+                    }
+                } else if (num_dispatchers == 4 || num_dispatchers == 8) {
+                    TT_ASSERT(device_worker_variants[DispatchWorkerType::DEMUX].size() == 3, "Insufficient Demux cores. Expected = 3. Found = {}", device_worker_variants[DispatchWorkerType::DEMUX].size());
+                    uint32_t dispatch_idx = 0;
+                    uint32_t demux_fanout = num_dispatchers / 2;
+                    for (int i = 1; i < 3; i++) {
+                        auto demux_settings = std::get<1>(device_worker_variants[DispatchWorkerType::DEMUX][i]);
```

Comment:
can we please add a comment explaining the topology here? I've scared that debug will be very painful for people who haven't brought up this feature, if they can't understand what's going on physically.

Reply:
I will update the tunneling diagram to show how N300 and Galaxy 1-2 CQ tunnels get connected.

Reply:
Thanks!